### ISSUE TITLE
Routing/lifecycle hooks

### ISSUE THREAD

**Issue by jgaskins** [clearwater-rb/clearwater#16]
It might not be a bad idea to add routing callbacks. This would allow for things like interrupting a transition or redirecting a transition due to authentication or authorization failure.

```ruby
class MyComponent
  include Clearwater::Component

  def will_transition_to(transition)
    unless authenticated?
      App.saved_url = transition.target_url # Save for after login
      transition.redirect("/login") # Replace target with "/login" and render that route instead
      alert "You must login before viewing that content."
    end
  end
end
```

Others for transitioning from or post-transition hooks might be helpful. I haven't used them in other frameworks, but they seem to be useful. I figure they could be implemented as no-ops on the `Component` mixin and the router calls them as part of every transition. This might involve moving the popstate handler into the router which, in hindsight, is probably a good idea anyway.

**Comment by jgaskins** [comment]
Thought about this a bit more on my way to work this morning. If we did this, it would probably only be added to components that serve as routing targets. Other components (such as those instantiated by the routing targets) wouldn't get this since the router can't know about them.

Another use case I thought about was displaying a Google Map. We would need to know when a component was rendered into the page so that we could trigger the map rendering, but we might not want to do that on every render. As far as I know, it's not currently possible to do this in Clearwater (unless someone's found a hack I don't know about yet). Similarly, we might want to know when we transition *away* from a particular route so we can tear down the state we built for that map.

Knowing when we transition away might also be good so that we can clear render caches for components that use the `CachedRender` mixin. `CachedRender` is great for allowing us to rerender a component frequently when it's already on the page and also allows us to hold onto it when we're pretty sure we're going to come back to it (for example, the most frequently used components in an app), but if we cache the render for a lot of different routes, we may be holding onto a lot of virtual-DOM nodes that might very well be stale the next time we render those components. This would give us a mechanism to clear those out when we navigate away.

**Comment by krainboltgreene** [comment]
So right now I'm running into an issue with being able to render a component and also trigger the initialization of `Chart.js`. Amusingly, reactjs does this pretty naturally: https://github.com/jhudson8/react-chartjs/blob/master/lib/core.js#L51-L57

I suspect we'll need more than just router-lifecycle points if I'm going to be able to do this. For reference here's all the reactjs lifecycle: https://facebook.github.io/react/docs/component-specs.html

**Comment by jgaskins** [comment]
> Amusingly, reactjs does this pretty naturally

**Comment by jgaskins** [comment]
> Amusingly, reactjs does this pretty naturally

Indeed. In React, the same location in the virtual DOM tree reuses the same component in memory as long as it's the same type of component. Clearwater doesn't "mount" components, though. We just use

In fact, we have no way of knowing what you're rendering other than routing targets, since all other components are instantiated within other components. I'm not even sure the `virtual-dom` library supports linking a VDOM node to a real DOM element. For handling things like form inputs, you have to do things like `event.target.value`.

Consider this component:

```ruby
class Foo
  include Clearwater::Component

  def render
    div(Bar.new)
  end
end
```

In this component, we instantiate a `Bar` component within the `Foo` component. Even if we route to a `Foo` instance, there is no way for to track the actual rendering of `Bar` (and especially _where_ it's rendered in the DOM) outside of `Foo` without doing it explicitly because we're just instantiating an object.

The reason I think routing callbacks would work for Chart.js or Google Maps or other libraries that require you to modify existing DOM nodes is because `Foo` _does_ know whether `Bar` will be rendered and (again, assuming it's a routing target) it can use its hook to tell the `Bar` component that it's been rendered into the DOM. Routing targets will have to be smarter than other components if we implement routing callbacks.

**Comment by jgaskins** [comment]
I think this problem will be solved by #34 & #35. That way, instead of lifecycle hooks, we can just use lightweight objects whose sole purpose is to represent a rendered DOM node or even completely take over the creation, modification, and teardown of a node.
